src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicIndexerAccessExpression.cs

```diff
@@ -0,0 +1,401 @@
+                // CS0631: ref and out are not valid in this context
+                //     public int this[ref int i, long c] => 0;
+                Diagnostic(ErrorCode.ERR_IllegalRefParam, "ref").WithLocation(11, 21)
+            };
+
+            VerifyOperationTreeAndDiagnosticsForTest<ElementAccessExpressionSyntax>(source, expectedOperationTree, expectedDiagnostics);
+        }
+
+        [CompilerTrait(CompilerFeature.IOperation)]
+        [Fact]
+        public void DynamicIndexerAccessExpression_ErrorBadDynamicMethodArgLambda()
+        {
+            string source = @"
+using System;
+
+class C
+{
+    public void M(C c)
+    {
+        dynamic y = null;
+        var x = /*<bind>*/c[delegate { }, y]/*</bind>*/;
+    }
+
+    public int this[Action a, Action y] => 0;
+}
+";
+            string expectedOperationTree = @"
+IDynamicIndexerAccessExpression (OperationKind.DynamicIndexerAccessExpression, Type: dynamic, IsInvalid) (Syntax: 'c[delegate { }, y]')
+  Expression: IParameterReferenceExpression: c (OperationKind.ParameterReferenceExpression, Type: C) (Syntax: 'c')
+  ApplicableSymbols(1):
+    Symbol: System.Int32 C.this[System.Action a, System.Action y] { get; }
+  Arguments(2):
+      IAnonymousFunctionExpression (Symbol: lambda expression) (OperationKind.AnonymousFunctionExpression, Type: null, IsInvalid) (Syntax: 'delegate { }')
+        IBlockStatement (1 statements) (OperationKind.BlockStatement, IsInvalid) (Syntax: '{ }')
+          IReturnStatement (OperationKind.ReturnStatement, IsInvalid) (Syntax: '{ }')
+            ReturnedValue: null
+      ILocalReferenceExpression: y (OperationKind.LocalReferenceExpression, Type: dynamic) (Syntax: 'y')
+  ArgumentNames(0)
+  ArgumentRefKinds(0)
+";
+            var expectedDiagnostics = new DiagnosticDescription[] {
+                // CS1977: Cannot use a lambda expression as an argument to a dynamically dispatched operation without first casting it to a delegate or expression tree type.
+                //         var x = /*<bind>*/c[delegate { }, y]/*</bind>*/;
+                Diagnostic(ErrorCode.ERR_BadDynamicMethodArgLambda, "delegate { }").WithLocation(9, 29)
+            };
+
+            VerifyOperationTreeAndDiagnosticsForTest<ElementAccessExpressionSyntax>(source, expectedOperationTree, expectedDiagnostics);
+        }
+
+        [CompilerTrait(CompilerFeature.IOperation)]
+        [Fact]
+        public void DynamicIndexerAccessExpression_OverloadResolutionFailure()
+        {
+            string source = @"
+using System;
+
+class C
+{
+    void M(C c, dynamic d)
+    {
+        var x = /*<bind>*/c[d]/*</bind>*/;
+    }
+
+    public int this[int i, int j] => 0;
+    public int this[int i, int j, int k] => 0;
+}
+";
+            string expectedOperationTree = @"
+IPropertyReferenceExpression: System.Int32 C.this { } (OperationKind.PropertyReferenceExpression, Type: System.Int32, IsInvalid) (Syntax: 'c[d]')
```

Comment:
>IPropertyReferenceExpression [](start = 0, length = 28)

I would expect an invalid expression here, similar to calls that fail overload resolution. Please open an issue to follow up.

Reply:
Opened https://github.com/dotnet/roslyn/issues/22043